### DISCUSSION TITLE
cognition.run external json stimuli

### DISCUSSION THREAD

**Discussion by AlexStuffs** [jspsych/jsPsych#1227] [Q&A] [answered]
Hi all, 

I'm working on a SART and an implementation onto cognition.run. 
I defined a list of stimuli in a json format file and it runs quite fine in jspsych.
However, I don't know how to load the stimuli in cognition.run. I uploaded the json file in cognition.run but I can't figure out how to pick up the stimuli in the timeline.
Do I have to use a for loop or something ?

Here the sample code about the stimuli files and timeline I used in cognition.run: 

```js
var  test_stimuli ="test_stimuli_cognition_short.json"
var number = {
      type: 'html-keyboard-response',
      stimulus: function() {
    // this is how you can combine different variables from your timeline_variables array
    return '<p style="font-size: '+jsPsych.timelineVariable('fontsize', true)+'px;">'+jsPsych.timelineVariable('number', true)+'</p>';
  },
      choices: ['space'],
      trial_duration: 250,
      data: {
        condition: jsPsych.timelineVariable('condition'),
      },
      on_finish: function(data){
        if(data.condition == 'Go'){
          data.correct = jsPsych.pluginAPI.compareKeys(data.key_press, 'space');
        } else {
          data.correct = jsPsych.pluginAPI.compareKeys(data.key_press, '');
        }
      }
    };
    var test = {
      timeline: [number, mask],
      timeline_variables: test_stimuli,mask,
      data: {
        'test_part': 'test'
      }
    }
```
My json file is like this: 
```js
[
  {
    "number": 9,
    "condition": "Go",
    "fontsize": 120
  },
  {
    "number": 6,
    "condition": "Go",
    "fontsize": 94
  },
  {
    "number": 1,
    "condition": "Go",
    "fontsize": 100
  },
  {
    "number": 4,
    "condition": "Go",
    "fontsize": 94
  },
]
```
Thanks!

Alex

**Comment by jodeleeuw** [comment, marked as answer]
Hi Alex,

It looks like there might be a typo here:

> `timeline_variables: test_stimuli,mask,`

The `mask` variable is probably a mistake?

Also, checkout [this video](https://www.youtube.com/watch?v=T3GCZWarEsY) about how to integrate JSON into an experiment. I recommend defining a variable name in the JSON file (making it a `.js` file) and then accessing that through the main experiment.

**Reply by AlexStuffs** [reply]
You're totally right! That's a typo that I have since corrected.

Thank your for the link, I was desperately looking for a tutorial or a cognition.run forum, I didn't know you made tutorials!
Thanks!!

**Comment by AlexStuffs** [comment]
My bad, I though that your video concerned the loading of external stimuli in cognition.run. 
I followed your tuto and make a stimuli.js file and it perfectly runs in classical way (ie without cognition.run).
However, when I implement it in cognition.run and upload the js file, the trial runs but stimuli are "undefined".

the undefined events are coded this ways: 
```
var number = {
  type: 'html-keyboard-response',
  stimulus: function() {
// this is how you can combine different variables from your timeline_variables array
return '<p style="font-size: '+jsPsych.timelineVariable('fontsize', true)+'px;">'+jsPsych.timelineVariable('number', true)+'</p>';
},
  choices: ['space'],
  trial_duration: 250,
  data: {
    condition: jsPsych.timelineVariable('condition'),
  },
  on_finish: function(data){
    if(data.condition == 'Go'){
      data.correct = jsPsych.pluginAPI.compareKeys(data.key_press, 'space');
    } else {
      data.correct = jsPsych.pluginAPI.compareKeys(data.key_press, '');
    }
  }
};
```
and the trial is run through: 
```
var test = {
  timeline: [number, mask],
  timeline_variables: [test_stimuli], <- cognition.run indicates here a referenceError: test_stimuli is not defined
  data: {
    'test_part': 'test'
  }
};
```
Moreover, the run stopped after 1 trial.

Alex

**Reply by jodeleeuw** [reply]
Hi Alex,

What does the .js file look like that you are uploading? I've used cognition.run and this exact strategy before, so I know that it can work in theory.

**Comment by AlexStuffs** [comment]
Edit: 
```
var test = {
  timeline: [number, mask],
  timeline_variables: [stimuli],
  data: {
    'test_part': 'test'
  }
};
```

My file stimuli.j looks like this: 
```
var stimuli = [
	  {
		number: 9,
		condition: "Go",
		fontsize: 120
	  },
	  {
		number: 6,
		condition: "Go",
		fontsize: 94
	  },
	  {
		number: 1,
		condition: "Go",
		fontsize: 100
	  },
	  {
		number: 4,
		condition: "Go",
		fontsize: 94
	  },
	  {
		number: 8,
		condition: "Go",
		fontsize: 72
	  },
	  {
		number: 6,
		condition: "Go",
		fontsize: 100
	  },
	  {
		number: 2,
		condition: "Go",
		fontsize: 100
	  },
]
```

**Reply by AlexStuffs** [reply]
And I upload it in the External JS/Css part

**Reply by jodeleeuw** [reply]
That sounds like it should work... can you post the link to the experiment?

**Reply by jodeleeuw** [reply]
Oh, I missed something. Since `stimuli` is already an array, you should not wrap it in `[]`:

```js
  timeline_variables: [stimuli],
``` 

should be

```js
  timeline_variables: stimuli,
```

**Reply by AlexStuffs** [reply]
I'm feeling soooooooooooo $*ù^$!:;$* 
I think I should take some js courses ^^'
Thank you!!

**Comment by AlexStuffs** [comment]
https://dn0eeub5qk.cognition.run/
